pkg/depend/graph.go

```diff
@@ -0,0 +1,180 @@
+)
+
+// Graph node only contains its direct dependencies.
+type Graph struct {
+	// TODO: Perhaps implement version here.
+	// Could be a sha of the dependency json.
+	// version    string
+	table      map[string]*Node
+	edges      []*Node
+	reposTable map[string]*repo
+}
+
+// Node of the tree.
+type Node struct {
+	name string
+	url  string
+	deps []*Node
+}
+
+type NodeCycleError struct {
+	nodeName    string
+	visitedName string
+}
+
+func (e *NodeCycleError) Error() string {
+	msg := fmt.Sprintf("Cycle detected.\n%s ---> %s ---> %s", e.visitedName, e.nodeName, e.visitedName)
+	return msg
+}
+
+// repo contains information about the repository
+// The direct dependencies in this struct are the names of other repos.
+type repo struct {
+	Name string   `json:"Name"`
+	URL  string   `json:"Url"`
+	Deps []string `json:"Deps,omitempty"`
+}
+
+// NewGraphFromFile unmarshalls JSON from a file into a graph.
+func NewGraphFromFile(path string) (*Graph, error) {
+	// Read and parse the data.
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	return NewGraph(data)
+}
+
+// NewGraph reads in the JSON data and creates the graph.
+func NewGraph(data []byte) (*Graph, error) {
+	var reps []*repo
+	if err := json.Unmarshal(data, &reps); err != nil {
+		return nil, err
+	}
+	// Collect all the entries and check for duplications.
+	repos := make(map[string]*repo)
+	for _, repo := range reps {
+		// Populate the table
+		if _, ok := repos[repo.Name]; !ok {
+			repos[repo.Name] = repo
+		} else {
+			return nil, errors.New("Duplicate key: " + repo.Name)
+		}
+	}
+	return newGraphFromRepos(repos)
+}
+
+// GetChildren returns a flat list of all children.
+// Does not contain duplicates.
+func (node *Node) GetChildren() []*Node {
+	visited := utils.NewSet()
+	var children []*Node
+	for _, d := range node.deps {
+		if !visited.Exists(d.name) {
+			visited.Add(d.name)
+			children = append(children, d)
+		}
+		for _, c := range d.GetChildren() {
+			if !visited.Exists(c.name) {
+				visited.Add(c.name)
+				children = append(children, c)
+			}
+		}
+	}
+	return children
+}
+
+func newGraphFromRepos(repos map[string]*repo) (*Graph, error) {
+	graph := &Graph{
+		table:      make(map[string]*Node, len(repos)),
+		reposTable: repos,
+	}
+
+	_, err := graph.resolve(graph.reposTable, nil)
+	if err != nil {
+		return nil, err
+	}
+	graph.createGraph()
+	return graph, nil
+}
+
+// resolveFromRepos will take the parsed JSON and resolve it into the Graph.Table
+func (graph *Graph) resolveFromRepos(repos map[string]*repo, visited *utils.StringSet) ([]*Node, error) {
+	for k, v := range repos {
+		for _, d := range v.Deps {
+			if visited.Exists(d) {
+				return nil, &NodeCycleError{k, d}
+			}
+		}
+	}
+	return graph.resolve(repos, visited)
+}
+
+func (graph *Graph) resolve(repos map[string]*repo, visited *utils.StringSet) ([]*Node, error) {
+	var top bool
+	if visited == nil {
+		top = true
+	}
+
+	var nodes []*Node
+	for k, v := range repos {
```

Comment:
Yes, we need it because of this line:

```
nod, err := graph.resolveFromRepos(deps, visited)
```
If you can think of a way to refactor this and it works with `go test`, let me know :)